pydicom/pixel_data_handlers/pillow_handler.py

```diff
@@ -162,16 +170,68 @@ def get_pixeldata(ds):
             im.draft('YCbCr', (ds.Rows, ds.Columns))
         pixel_bytes.extend(im.tobytes())
 
+        j2k_precision = _get_j2k_precision(pixel_data)
+
     logger.debug("Successfully read %s pixel bytes", len(pixel_bytes))
 
     arr = numpy.frombuffer(pixel_bytes, pixel_dtype(ds))
 
-    if (transfer_syntax in PillowJPEG2000TransferSyntaxes and
-            ds.BitsStored == 16):
-        # WHY IS THIS EVEN NECESSARY??
-        arr &= 0x7FFF
+    if transfer_syntax in PillowJPEG2000TransferSyntaxes:
+        # See #693 for justification of flipping MSB and bit shifting
+        if ds.BitsAllocated == 16 and ds.PixelRepresentation == 1:
+            # WHY IS THIS EVEN NECESSARY??
+            # Flip MSb: b10000000 00000000
+            arr ^= 0x8000
+
+        if j2k_precision and j2k_precision != ds.BitsStored:
+            warnings.warn(
```

Comment:
I went with this rather than automatically updating `ds.BitsStored` using `j2k_precision` because I wanted the user to be able to control things rather than having a "solution" forced upon then that may end up making things worse.

I think under almost all situations the `j2k_precision` will be the correct one, though.

Reply:
Hm, I understand the sentiment, though I think we should do the correction in this case (with an additional warning), as it is almost certainly the right thing to do. Most pydicom users just want to be able to access the correct pixel data, and are not concerned about fixing incorrect DICOM coming from elsewhere. If something is wrong, there is always the warning that states the reason (and the probable fix) for that. I may be wrong, of course - going to wait for @darcymason's opinion. 

By the way: awesome work! Even it is still unclear why this is needed (I haven't dug into this myself), it seems to fix this longstanding issue.

Reply:
> going to wait for @darcymason's opinion

What happens downstream here?  It seems like this won't affect anything, except if user were to write a DICOM file, it would continue to have an incorrected BitsStored value?  Am I interpreting that correctly?

Reply:
Yes, as far as I understand, and I'm fine with that (especially after the later clarification and change in the latest commit - see below).